src/Core/ConstraintValidator/ZipCodeRangeValidator.php

```diff
@@ -0,0 +1,124 @@
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Core\ConstraintValidator;
+
+use PrestaShop\PrestaShop\Core\CommandBus\CommandBusInterface;
+use PrestaShop\PrestaShop\Core\Domain\Country\Exception\CountryConstraintException;
+use PrestaShop\PrestaShop\Core\Domain\Country\Query\GetCountryZipCodeRequirements;
+use PrestaShop\PrestaShop\Core\Domain\Country\QueryResult\CountryZipCodeRequirements;
+use PrestaShop\PrestaShop\Core\Domain\Country\ValueObject\CountryId;
+use PrestaShop\PrestaShop\Core\Form\FormChoiceProviderInterface;
+use Symfony\Component\Translation\TranslatorInterface;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+
+/**
+ * Validate zip codes in range
+ */
+class ZipCodeRangeValidator extends ConstraintValidator
+{
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    /**
+     * @var CommandBusInterface
+     */
+    private $queryBus;
+
+    /**
+     * @var FormChoiceProviderInterface
+     */
+    private $countryChoiceProvider;
+
+    /**
+     * @param TranslatorInterface $translator
+     * @param CommandBusInterface $queryBus
+     * @param FormChoiceProviderInterface $countryChoiceProvider
+     */
+    public function __construct(
+        TranslatorInterface $translator,
+        CommandBusInterface $queryBus,
+        FormChoiceProviderInterface $countryChoiceProvider
+    ) {
+        $this->translator = $translator;
+        $this->queryBus = $queryBus;
+        $this->countryChoiceProvider = $countryChoiceProvider;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws CountryConstraintException
+     */
+    public function validate($value, Constraint $constraint)
+    {
+        $countryId = $value['country_id'] ?? CountryId::ALL_COUNTRIES_ID;
+
+        $selectedCountries = [$countryId];
+        $zipCode = $value['zip_code'];
+
+        if (null === $zipCode) {
+            return;
+        }
+
+        if ($countryId === 0) {
+            $selectedCountries = array_map('intval', $this->countryChoiceProvider->getChoices());
+        }
+
+        foreach ($selectedCountries as $country) {
+            /** @var CountryZipCodeRequirements $requirements */
+            $requirements = $this->queryBus->handle(new GetCountryZipCodeRequirements($country));
+
+            foreach (explode('-', $zipCode) as $code) {
+                $isInvalid = (bool) $code &&
+                    null !== $requirements->getPattern() &&
+                    !(bool) preg_match($requirements->getPattern(), trim($code));
+
+                if ($isInvalid) {
+                    $message = $this->translator->trans(
+                        'The Zip/postal code is invalid. It must be typed as follows: %format% for %country%.',
```

Comment:
This string is located in Admin.International.Notification, right?

Reply:
I've just noticed the domain by default is Admin.international.Notification anyway ;)